File: src/app-aft/wp-fold.js
function wp_fold(x)
{
 check(is_arr,x)
 
 function get_children(x,y)
 {
  check(is_arr,x)
  check(is_obj,y)
  
  const r=arr()
  
  for(const v of x)
  {
   if(different(v.parent,y.id))
    continue
    
   const o=dup(v)
   
   o.children=get_children(x,o)
   
   push(r,o)
  }
  
  return r
 }
 
 const r=arr()
 const a=dup(x)
 
 while(is_full(a))
 {
  const o=shift(a)
  
  if(same(o.parent,0))
  {
   o.children=get_children(x,o)
  
   push(r,o)
  }
 }
 
 return r
}

File: src/lib-aft/wp-fold.cs
fn wp_fold x:arr
 check is_arr x

 fn get_children x:arr y:obj
  //~ check is_arr x
  //~ check is_obj y

  let r arr

  forof x
   if different v.parent y.id
    cont

   let o dup v

   assign o.children get_children x o

   push r o
  end

  ret r
 end

 let r arr
 let a dup x

 while is_full a
  let o shift a

  if same o.parent 0
   assign o.children get_children x o

   push r o
  end
 end

 ret r
end
